From: Alaa Hleihel <alaa@mellanox.com>
Subject: [PATCH] BACKPORT: ib_srp

Add only ib_srp backports to this patch.
That is:
- drivers/infiniband/ulp/srp/*
- drivers/scsi/scsi_priv.h

Change-Id: I80c35a953cb9a3f9e7878211faaa5f4fad007889
Signed-off-by: Alaa Hleihel <alaa@mellanox.com>
---
 drivers/infiniband/ulp/srp/ib_srp.c | 317 +++++++++++++++++++++++++++++++++++-
 drivers/infiniband/ulp/srp/ib_srp.h |  29 ++++
 drivers/scsi/scsi_priv.h            |   5 +
 drivers/scsi/scsi_transport_srp.c   |  21 +++
 4 files changed, 371 insertions(+), 1 deletion(-)

--- a/drivers/infiniband/ulp/srp/ib_srp.c
+++ b/drivers/infiniband/ulp/srp/ib_srp.c
@@ -73,6 +73,13 @@ MODULE_INFO(release_date, DRV_RELDATE);
 #define DYNAMIC_DEBUG_BRANCH(descriptor) false
 #endif
 
+#ifndef DEFINE_DYNAMIC_DEBUG_METADATA
+#define DEFINE_DYNAMIC_DEBUG_METADATA(name, fmt)
+#endif
+#ifndef DYNAMIC_DEBUG_BRANCH
+#define DYNAMIC_DEBUG_BRANCH(descriptor) false
+#endif
+
 static unsigned int srp_sg_tablesize;
 static unsigned int cmd_sg_entries;
 static unsigned int indirect_sg_entries;
@@ -86,12 +93,22 @@ module_param(srp_sg_tablesize, uint, 044
 MODULE_PARM_DESC(srp_sg_tablesize, "Deprecated name for cmd_sg_entries");
 
 module_param(cmd_sg_entries, uint, 0444);
+#ifdef HAVE_BLK_QUEUE_VIRT_BOUNDARY
 MODULE_PARM_DESC(cmd_sg_entries,
 		 "Default number of gather/scatter entries in the SRP command (default is 12, max 255)");
+#else
+MODULE_PARM_DESC(cmd_sg_entries,
+                 "Default number of gather/scatter entries in the SRP command (default is 12, max 12)");
+#endif
 
 module_param(indirect_sg_entries, uint, 0444);
+#ifdef HAVE_SG_MAX_SEGMENTS
 MODULE_PARM_DESC(indirect_sg_entries,
 		 "Default max number of gather/scatter entries (default is 12, max is " __stringify(SG_MAX_SEGMENTS) ")");
+#else
+MODULE_PARM_DESC(indirect_sg_entries,
+		 "Default max number of gather/scatter entries (default is 12, max is " __stringify(SCSI_MAX_SG_CHAIN_SEGMENTS) ")");
+#endif
 
 module_param(allow_ext_sg, bool, 0444);
 MODULE_PARM_DESC(allow_ext_sg,
@@ -112,6 +129,7 @@ MODULE_PARM_DESC(register_always,
 module_param(never_register, bool, 0444);
 MODULE_PARM_DESC(never_register, "Never register memory");
 
+#ifdef HAVE_MODULEPARAM_KERNEL_PARAM_OPS
 static const struct kernel_param_ops srp_tmo_ops;
 
 static int srp_reconnect_delay = 20;
@@ -130,6 +148,30 @@ MODULE_PARM_DESC(fast_io_fail_tmo,
 static int srp_dev_loss_tmo = 600;
 module_param_cb(dev_loss_tmo, &srp_tmo_ops, &srp_dev_loss_tmo,
 		S_IRUGO | S_IWUSR);
+#else
+static int srp_tmo_get(char *buffer, const struct kernel_param *kp);
+static int srp_tmo_set(const char *val, const struct kernel_param *kp);
+
+static int srp_reconnect_delay = 20;
+module_param_call(reconnect_delay, (param_set_fn)srp_tmo_set,
+		  (param_get_fn)srp_tmo_get, &srp_reconnect_delay,
+		  S_IRUGO | S_IWUSR);
+MODULE_PARM_DESC(reconnect_delay, "Time between successive reconnect attempts");
+
+static int srp_fast_io_fail_tmo = 15;
+module_param_call(fast_io_fail_tmo, (param_set_fn)srp_tmo_set,
+		  (param_get_fn)srp_tmo_get, &srp_fast_io_fail_tmo,
+		  S_IRUGO | S_IWUSR);
+MODULE_PARM_DESC(fast_io_fail_tmo,
+		 "Number of seconds between the observation of a transport"
+		 " layer error and failing all I/O. \"off\" means that this"
+		 " functionality is disabled.");
+
+static int srp_dev_loss_tmo = 600;
+module_param_call(dev_loss_tmo, (param_set_fn)srp_tmo_set,
+		  (param_get_fn)srp_tmo_get, &srp_dev_loss_tmo,
+		  S_IRUGO | S_IWUSR);
+#endif
 MODULE_PARM_DESC(dev_loss_tmo,
 		 "Maximum number of seconds that the SRP transport should"
 		 " insulate transport layer errors. After this time has been"
@@ -195,10 +237,12 @@ out:
 	return res;
 }
 
+#ifdef HAVE_MODULEPARAM_KERNEL_PARAM_OPS
 static const struct kernel_param_ops srp_tmo_ops = {
 	.get = srp_tmo_get,
 	.set = srp_tmo_set,
 };
+#endif
 
 static inline struct srp_target_port *host_to_target(struct Scsi_Host *host)
 {
@@ -867,6 +911,9 @@ static int srp_alloc_req_data(struct srp
 	dma_addr_t dma_addr;
 	int i, ret = -ENOMEM;
 
+#ifndef HAVE_BLK_MQ_UNIQUE_TAG
+	INIT_LIST_HEAD(&ch->free_reqs);
+#endif
 	ch->req_ring = kcalloc(target->req_ring_size, sizeof(*ch->req_ring),
 			       GFP_KERNEL);
 	if (!ch->req_ring)
@@ -898,6 +945,10 @@ static int srp_alloc_req_data(struct srp
 			goto out;
 
 		req->indirect_dma_addr = dma_addr;
+#ifndef HAVE_BLK_MQ_UNIQUE_TAG
+		req->tag = build_srp_tag(ch - target->ch, i);
+		list_add_tail(&req->list, &ch->free_reqs);
+#endif
 	}
 	ret = 0;
 
@@ -943,6 +994,10 @@ static void srp_remove_target(struct srp
 		ch = &target->ch[i];
 		srp_free_req_data(target, ch);
 	}
+#ifndef HAVE_BLK_MQ_UNIQUE_TAG
+	kfree(target->mq_map);
+	target->mq_map = NULL;
+#endif
 	kfree(target->ch);
 	target->ch = NULL;
 
@@ -1151,6 +1206,9 @@ static void srp_free_req(struct srp_rdma
 
 	spin_lock_irqsave(&ch->lock, flags);
 	ch->req_lim += req_lim_delta;
+#ifndef HAVE_BLK_MQ_UNIQUE_TAG
+	list_add_tail(&req->list, &ch->free_reqs);
+#endif
 	spin_unlock_irqrestore(&ch->lock, flags);
 }
 
@@ -1170,16 +1228,20 @@ static void srp_terminate_io(struct srp_
 {
 	struct srp_target_port *target = rport->lld_data;
 	struct srp_rdma_ch *ch;
+#ifdef HAVE_REQUEST_QUEUE_REQUEST_FN_ACTIVE
 	struct Scsi_Host *shost = target->scsi_host;
 	struct scsi_device *sdev;
+#endif
 	int i, j;
 
 	/*
 	 * Invoking srp_terminate_io() while srp_queuecommand() is running
 	 * is not safe. Hence the warning statement below.
 	 */
+#ifdef HAVE_REQUEST_QUEUE_REQUEST_FN_ACTIVE
 	shost_for_each_device(sdev, shost)
 		WARN_ON_ONCE(sdev->request_queue->request_fn_active);
+#endif
 
 	for (i = 0; i < target->ch_count; i++) {
 		ch = &target->ch[i];
@@ -1902,6 +1964,9 @@ static void srp_process_rsp(struct srp_r
 	struct srp_request *req;
 	struct scsi_cmnd *scmnd;
 	unsigned long flags;
+#ifndef HAVE_BLK_MQ_UNIQUE_TAG
+	unsigned i;
+#endif
 
 	if (unlikely(rsp->tag & SRP_TAG_TSK_MGMT)) {
 		spin_lock_irqsave(&ch->lock, flags);
@@ -1918,6 +1983,7 @@ static void srp_process_rsp(struct srp_r
 		}
 		spin_unlock_irqrestore(&ch->lock, flags);
 	} else {
+#ifdef HAVE_BLK_MQ_UNIQUE_TAG
 		scmnd = scsi_host_find_tag(target->scsi_host, rsp->tag);
 		if (scmnd && scmnd->host_scribble) {
 			req = (void *)scmnd->host_scribble;
@@ -1925,6 +1991,18 @@ static void srp_process_rsp(struct srp_r
 		} else {
 			scmnd = NULL;
 		}
+#else
+		if (srp_tag_ch(rsp->tag) != ch - target->ch)
+			pr_err("Channel idx mismatch: tag %#llx <> ch %#lx\n",
+				rsp->tag, ch - target->ch);
+		i = srp_tag_idx(rsp->tag);
+		if (i < target->req_ring_size) {
+			req = &ch->req_ring[i];
+			scmnd = srp_claim_req(ch, req, NULL, NULL);
+		} else {
+			scmnd = NULL;
+		}
+#endif
 		if (!scmnd) {
 			shost_printk(KERN_ERR, target->scsi_host,
 				     "Null scmnd for RSP w/tag %#016llx received on ch %td / QP %#x\n",
@@ -2020,8 +2098,13 @@ static void srp_process_aer_req(struct s
 	};
 	s32 delta = be32_to_cpu(req->req_lim_delta);
 
+#ifdef HAVE_SCSI_DEVICE_U64_LUN
 	shost_printk(KERN_ERR, target->scsi_host, PFX
 		     "ignoring AER for LUN %llu\n", scsilun_to_int(&req->lun));
+#else
+	shost_printk(KERN_ERR, target->scsi_host, PFX
+		     "ignoring AER for LUN %u\n", scsilun_to_int(&req->lun));
+#endif
 
 	if (srp_response_common(ch, delta, &rsp, sizeof(rsp)))
 		shost_printk(KERN_ERR, target->scsi_host, PFX
@@ -2120,7 +2203,57 @@ static void srp_handle_qp_err(struct ib_
 	target->qp_in_error = true;
 }
 
-static int srp_queuecommand(struct Scsi_Host *shost, struct scsi_cmnd *scmnd)
+#ifndef HAVE_BLK_MQ_UNIQUE_TAG
+static struct srp_rdma_ch *srp_map_cpu_to_ch(struct srp_target_port *target)
+{
+	return &target->ch[target->mq_map[raw_smp_processor_id()]];
+}
+#endif
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 36)
+/*
+ * Kernel with host lock push-down patch. See also upstream commit
+ * f281233d3eba15fb225d21ae2e228fd4553d824a.
+ */
+#define SRP_QUEUECOMMAND srp_queuecommand
+#elif defined(RHEL_MAJOR) && RHEL_MAJOR -0 == 6 && RHEL_MINOR -0 >= 2
+/*
+ * Kernel with lockless SCSI command dispatching enabled.
+ * See also the RHEL 6.2 release notes (http://access.redhat.com/knowledge/docs/en-US/Red_Hat_Enterprise_Linux/6/html-single/6.2_Release_Notes/index.html).
+ */
+static int srp_queuecommand_wrk(struct Scsi_Host *shost,
+				struct scsi_cmnd *scmnd);
+static int srp_queuecommand(struct scsi_cmnd *scmnd,
+			    void (*done)(struct scsi_cmnd *))
+{
+	scmnd->scsi_done = done;
+	return srp_queuecommand_wrk(scmnd->device->host, scmnd);
+}
+#define SRP_QUEUECOMMAND srp_queuecommand_wrk
+#else
+/*
+ * Kernel that invokes srp_queuecommand with the SCSI host lock held.
+ */
+static int srp_queuecommand_wrk(struct Scsi_Host *shost,
+				struct scsi_cmnd *scmnd);
+static int srp_queuecommand(struct scsi_cmnd *scmnd,
+			    void (*done)(struct scsi_cmnd *))
+{
+	struct Scsi_Host *shost = scmnd->device->host;
+	int res;
+
+	spin_unlock_irq(shost->host_lock);
+
+	scmnd->scsi_done = done;
+	res = srp_queuecommand_wrk(shost, scmnd);
+
+	spin_lock_irq(shost->host_lock);
+	return res;
+}
+#define SRP_QUEUECOMMAND srp_queuecommand_wrk
+#endif
+
+static int SRP_QUEUECOMMAND(struct Scsi_Host *shost, struct scsi_cmnd *scmnd)
 {
 	struct srp_target_port *target = host_to_target(shost);
 	struct srp_rport *rport = target->rport;
@@ -2131,7 +2264,9 @@ static int srp_queuecommand(struct Scsi_
 	struct ib_device *dev;
 	unsigned long flags;
 	u32 tag;
+#ifdef HAVE_BLK_MQ_UNIQUE_TAG
 	u16 idx;
+#endif
 	int len, ret;
 	const bool in_scsi_eh = !in_interrupt() && current == shost->ehandler;
 
@@ -2148,6 +2283,7 @@ static int srp_queuecommand(struct Scsi_
 	if (unlikely(scmnd->result))
 		goto err;
 
+#ifdef HAVE_BLK_MQ_UNIQUE_TAG
 	WARN_ON_ONCE(scmnd->request->tag < 0);
 	tag = blk_mq_unique_tag(scmnd->request);
 	ch = &target->ch[blk_mq_unique_tag_to_hwq(tag)];
@@ -2155,15 +2291,28 @@ static int srp_queuecommand(struct Scsi_
 	WARN_ONCE(idx >= target->req_ring_size, "%s: tag %#x: idx %d >= %d\n",
 		  dev_name(&shost->shost_gendev), tag, idx,
 		  target->req_ring_size);
+#else
+	ch = srp_map_cpu_to_ch(target);
+#endif
 
 	spin_lock_irqsave(&ch->lock, flags);
 	iu = __srp_get_tx_iu(ch, SRP_IU_CMD);
+#ifdef HAVE_BLK_MQ_UNIQUE_TAG
 	spin_unlock_irqrestore(&ch->lock, flags);
 
 	if (!iu)
 		goto err;
 
 	req = &ch->req_ring[idx];
+#else
+	if (!iu)
+		goto err_unlock;
+
+	req = list_first_entry(&ch->free_reqs, struct srp_request, list);
+	list_del(&req->list);
+	tag = req->tag;
+	spin_unlock_irqrestore(&ch->lock, flags);
+#endif
 	dev = target->srp_host->srp_dev->dev;
 	ib_dma_sync_single_for_cpu(dev, iu->dma, target->max_iu_len,
 				   DMA_TO_DEVICE);
@@ -2224,6 +2373,14 @@ err_iu:
 	 */
 	req->scmnd = NULL;
 
+#ifndef HAVE_BLK_MQ_UNIQUE_TAG
+	spin_lock_irqsave(&ch->lock, flags);
+	list_add(&req->list, &ch->free_reqs);
+
+err_unlock:
+	spin_unlock_irqrestore(&ch->lock, flags);
+
+#endif
 err:
 	if (scmnd->result) {
 		scmnd->scsi_done(scmnd);
@@ -2539,6 +2696,31 @@ static int srp_cm_handler(struct ib_cm_i
 	return 0;
 }
 
+#if defined(HAVE_SCSI_HOST_TEMPLATE_CHANGE_QUEUE_TYPE) && \
+	!defined(HAVE_SCSI_TCQ_SCSI_CHANGE_QUEUE_TYPE)
+/**
+ * srp_change_queue_type - changing device queue tag type
+ * @sdev: scsi device struct
+ * @tag_type: requested tag type
+ *
+ * Returns queue tag type.
+ */
+static int
+srp_change_queue_type(struct scsi_device *sdev, int tag_type)
+{
+	if (sdev->tagged_supported) {
+		scsi_set_tag_type(sdev, tag_type);
+		if (tag_type)
+			scsi_activate_tcq(sdev, sdev->queue_depth);
+		else
+			scsi_deactivate_tcq(sdev, sdev->queue_depth);
+	} else
+		tag_type = 0;
+
+	return tag_type;
+}
+#endif
+
 /**
  * srp_change_queue_depth - setting device queue depth
  * @sdev: scsi device struct
@@ -2546,13 +2728,40 @@ static int srp_cm_handler(struct ib_cm_i
  *
  * Returns queue depth.
  */
+#ifdef HAVE_SCSI_HOST_TEMPLATE_TRACK_QUEUE_DEPTH
 static int
 srp_change_queue_depth(struct scsi_device *sdev, int qdepth)
 {
 	if (!sdev->tagged_supported)
 		qdepth = 1;
+#ifdef HAVE_SCSI_CHANGE_QUEUE_DEPTH
 	return scsi_change_queue_depth(sdev, qdepth);
+#else
+	scsi_adjust_queue_depth(sdev, qdepth);
+	return sdev->queue_depth;
+#endif //HAVE_SCSI_CHANGE_QUEUE_DEPTH
 }
+#else
+static int
+srp_change_queue_depth(struct scsi_device *sdev, int qdepth, int reason)
+{
+	struct Scsi_Host *shost = sdev->host;
+	int max_depth;
+	if (reason == SCSI_QDEPTH_DEFAULT || reason == SCSI_QDEPTH_RAMP_UP) {
+		max_depth = shost->can_queue;
+		if (!sdev->tagged_supported)
+			max_depth = 1;
+		if (qdepth > max_depth)
+			qdepth = max_depth;
+		scsi_adjust_queue_depth(sdev, scsi_get_tag_type(sdev), qdepth);
+	} else if (reason == SCSI_QDEPTH_QFULL)
+		scsi_track_queue_full(sdev, qdepth);
+	else
+		return -EOPNOTSUPP;
+
+	return sdev->queue_depth;
+}
+#endif //HAVE_SCSI_HOST_TEMPLATE_TRACK_QUEUE_DEPTH
 
 static int srp_send_tsk_mgmt(struct srp_rdma_ch *ch, u64 req_tag, u64 lun,
 			     u8 func, u8 *status)
@@ -2631,8 +2840,13 @@ static int srp_abort(struct scsi_cmnd *s
 
 	if (!req)
 		return SUCCESS;
+#ifdef HAVE_BLK_MQ_UNIQUE_TAG
 	tag = blk_mq_unique_tag(scmnd->request);
 	ch_idx = blk_mq_unique_tag_to_hwq(tag);
+#else
+	tag = req->tag;
+	ch_idx = srp_tag_ch(tag);
+#endif
 	if (WARN_ON_ONCE(ch_idx >= target->ch_count))
 		return SUCCESS;
 	ch = &target->ch[ch_idx];
@@ -2691,6 +2905,7 @@ static int srp_reset_host(struct scsi_cm
 	return srp_reconnect_rport(target->rport) == 0 ? SUCCESS : FAILED;
 }
 
+#ifdef HAVE_BLK_QUEUE_VIRT_BOUNDARY
 static int srp_slave_alloc(struct scsi_device *sdev)
 {
 	struct Scsi_Host *shost = sdev->host;
@@ -2704,6 +2919,7 @@ static int srp_slave_alloc(struct scsi_d
 
 	return 0;
 }
+#endif
 
 static int srp_slave_configure(struct scsi_device *sdev)
 {
@@ -2896,11 +3112,23 @@ static struct scsi_host_template srp_tem
 	.module				= THIS_MODULE,
 	.name				= "InfiniBand SRP initiator",
 	.proc_name			= DRV_NAME,
+#ifdef HAVE_BLK_QUEUE_VIRT_BOUNDARY
 	.slave_alloc			= srp_slave_alloc,
+#endif
 	.slave_configure		= srp_slave_configure,
 	.info				= srp_target_info,
+#ifdef HAVE_SCSI_HOST_TEMPLATE_LOCKLESS
+	.lockless                       = true,
+#endif
 	.queuecommand			= srp_queuecommand,
 	.change_queue_depth             = srp_change_queue_depth,
+#ifdef HAVE_SCSI_HOST_TEMPLATE_CHANGE_QUEUE_TYPE
+#ifdef HAVE_SCSI_TCQ_SCSI_CHANGE_QUEUE_TYPE
+	.change_queue_type		= scsi_change_queue_type,
+#else
+	.change_queue_type		= srp_change_queue_type,
+#endif
+#endif
 	.eh_timed_out			= srp_timed_out,
 	.eh_abort_handler		= srp_abort,
 	.eh_device_reset_handler	= srp_reset_device,
@@ -2912,7 +3140,15 @@ static struct scsi_host_template srp_tem
 	.cmd_per_lun			= SRP_DEFAULT_CMD_SQ_SIZE,
 	.use_clustering			= ENABLE_CLUSTERING,
 	.shost_attrs			= srp_host_attrs,
+#ifdef HAVE_SCSI_HOST_TEMPLATE_USE_HOST_WIDE_TAGS
+	.use_host_wide_tags		= 1,
+#endif
+#ifdef HAVE_SCSI_HOST_TEMPLATE_USE_BLK_TAGS
+	.use_blk_tags			= 1,
+#endif
+#ifdef HAVE_SCSI_HOST_TEMPLATE_TRACK_QUEUE_DEPTH
 	.track_queue_depth		= 1,
+#endif
 };
 
 static int srp_sdev_count(struct Scsi_Host *host)
@@ -2961,8 +3197,13 @@ static int srp_add_target(struct srp_hos
 	list_add_tail(&target->list, &host->target_list);
 	spin_unlock(&host->target_lock);
 
+#ifdef HAVE_SCSI_SCAN_INITIAL
 	scsi_scan_target(&target->scsi_host->shost_gendev,
 			 0, target->scsi_id, SCAN_WILD_CARD, SCSI_SCAN_INITIAL);
+#else
+        scsi_scan_target(&target->scsi_host->shost_gendev,
+                         0, target->scsi_id, SCAN_WILD_CARD, 0);
+#endif
 
 	if (srp_connected_ch(target) < target->ch_count ||
 	    target->qp_in_error) {
@@ -3222,11 +3463,19 @@ static int srp_parse_options(const char
 			break;
 
 		case SRP_OPT_CMD_SG_ENTRIES:
+#ifdef HAVE_BLK_QUEUE_VIRT_BOUNDARY
 			if (match_int(args, &token) || token < 1 || token > 255) {
 				pr_warn("bad max cmd_sg_entries parameter '%s'\n",
 					p);
 				goto out;
 			}
+#else
+			if (match_int(args, &token) || token < 1 || token > 12) {
+				pr_warn("bad max cmd_sg_entries parameter '%s'\n",
+					p);
+				goto out;
+			}
+#endif
 			target->cmd_sg_cnt = token;
 			break;
 
@@ -3239,12 +3488,21 @@ static int srp_parse_options(const char
 			break;
 
 		case SRP_OPT_SG_TABLESIZE:
+#ifdef HAVE_SG_MAX_SEGMENTS
 			if (match_int(args, &token) || token < 1 ||
 					token > SG_MAX_SEGMENTS) {
 				pr_warn("bad max sg_tablesize parameter '%s'\n",
 					p);
 				goto out;
 			}
+#else
+			if (match_int(args, &token) || token < 1 ||
+					token > SCSI_MAX_SG_CHAIN_SEGMENTS) {
+				pr_warn("bad max sg_tablesize parameter '%s'\n",
+					p);
+				goto out;
+			}
+#endif
 			target->sg_tablesize = token;
 			break;
 
@@ -3304,7 +3562,11 @@ static ssize_t srp_create_target(struct
 	struct srp_device *srp_dev = host->srp_dev;
 	struct ib_device *ibdev = srp_dev->dev;
 	int ret, node_idx, node, cpu, i;
+#ifdef HAVE_BLK_QUEUE_VIRT_BOUNDARY
 	unsigned int max_sectors_per_mr, mr_per_cmd = 0;
+#else
+	unsigned int mr_per_cmd = 0;
+#endif
 	bool multich = false;
 
 	target_host = scsi_host_alloc(&srp_template,
@@ -3312,6 +3574,11 @@ static ssize_t srp_create_target(struct
 	if (!target_host)
 		return -ENOMEM;
 
+#if defined(HAVE_SCSI_HOST_USE_BLK_MQ) && \
+	defined(HAVE_SCSI_TCQ_SCSI_INIT_SHARED_TAG_MAP) && \
+	!defined(HAVE_SCSI_HOST_TEMPLATE_USE_BLK_TAGS)
+	target_host->use_blk_mq = true;
+#endif
 	target_host->transportt  = ib_srp_transport_template;
 	target_host->max_channel = 0;
 	target_host->max_id      = 1;
@@ -3345,6 +3612,12 @@ static ssize_t srp_create_target(struct
 	if (ret)
 		goto out;
 
+#ifdef HAVE_SCSI_HOST_TEMPLATE_USE_BLK_TAGS
+	ret = scsi_init_shared_tag_map(target_host, target_host->can_queue);
+	if (ret)
+		goto out;
+#endif
+
 	target->req_ring_size = target->queue_size - SRP_TSK_MGMT_SQ_SIZE;
 
 	if (!srp_conn_unique(target->srp_host, target)) {
@@ -3367,6 +3640,7 @@ static ssize_t srp_create_target(struct
 		bool gaps_reg = (ibdev->attrs.device_cap_flags &
 				 IB_DEVICE_SG_GAPS_REG);
 
+#ifdef HAVE_BLK_QUEUE_VIRT_BOUNDARY
 		max_sectors_per_mr = srp_dev->max_pages_per_mr <<
 				  (ilog2(srp_dev->mr_page_size) - 9);
 		if (!gaps_reg) {
@@ -3394,6 +3668,12 @@ static ssize_t srp_create_target(struct
 			 gaps_reg, target->scsi_host->max_sectors,
 			 srp_dev->max_pages_per_mr, srp_dev->mr_page_size,
 			 max_sectors_per_mr, mr_per_cmd);
+#else
+		if (!gaps_reg)
+			mr_per_cmd = target->cmd_sg_cnt;
+		else
+			mr_per_cmd = 1;
+#endif
 	}
 
 	target_host->sg_tablesize = target->sg_tablesize;
@@ -3423,6 +3703,12 @@ static ssize_t srp_create_target(struct
 	if (!target->ch)
 		goto out;
 
+#ifndef HAVE_BLK_MQ_UNIQUE_TAG
+	target->mq_map = kcalloc(nr_cpu_ids, sizeof(*target->mq_map),
+				 GFP_KERNEL);
+	if (!target->mq_map)
+		goto err_free_ch;
+#endif
 	node_idx = 0;
 	for_each_online_node(node) {
 		const int ch_start = (node_idx * target->ch_count /
@@ -3440,6 +3726,10 @@ static ssize_t srp_create_target(struct
 		for_each_online_cpu(cpu) {
 			if (cpu_to_node(cpu) != node)
 				continue;
+#ifndef HAVE_BLK_MQ_UNIQUE_TAG
+			target->mq_map[cpu] = ch_start == ch_end ? ch_start :
+				ch_start + cpu_idx % (ch_end - ch_start);
+#endif
 			if (ch_start + cpu_idx >= ch_end)
 				continue;
 			ch = &target->ch[ch_start + cpu_idx];
@@ -3472,6 +3762,9 @@ static ssize_t srp_create_target(struct
 				} else {
 					srp_free_ch_ib(target, ch);
 					srp_free_req_data(target, ch);
+#ifndef HAVE_BLK_MQ_UNIQUE_TAG
+					target->mq_map[cpu] = 0;
+#endif
 					target->ch_count = ch - target->ch;
 					goto connected;
 				}
@@ -3484,7 +3777,9 @@ static ssize_t srp_create_target(struct
 	}
 
 connected:
+#ifdef HAVE_SCSI_HOST_NR_HW_QUEUES
 	target->scsi_host->nr_hw_queues = target->ch_count;
+#endif
 
 	ret = srp_add_target(host, target);
 	if (ret)
@@ -3522,6 +3817,11 @@ free_ch:
 		srp_free_req_data(target, ch);
 	}
 
+#ifndef HAVE_BLK_MQ_UNIQUE_TAG
+	kfree(target->mq_map);
+
+err_free_ch:
+#endif
 	kfree(target->ch);
 	goto out;
 }
@@ -3730,10 +4030,17 @@ static int __init srp_init_module(void)
 	if (!cmd_sg_entries)
 		cmd_sg_entries = SRP_DEF_SG_TABLESIZE;
 
+#ifdef HAVE_BLK_QUEUE_VIRT_BOUNDARY
 	if (cmd_sg_entries > 255) {
 		pr_warn("Clamping cmd_sg_entries to 255\n");
 		cmd_sg_entries = 255;
 	}
+#else
+	if (cmd_sg_entries > 12) {
+		pr_warn("Clamping cmd_sg_entries to 12\n");
+		cmd_sg_entries = 12;
+	}
+#endif
 
 	if (!indirect_sg_entries)
 		indirect_sg_entries = cmd_sg_entries;
@@ -3743,11 +4050,19 @@ static int __init srp_init_module(void)
 		indirect_sg_entries = cmd_sg_entries;
 	}
 
+#ifdef HAVE_SG_MAX_SEGMENTS
 	if (indirect_sg_entries > SG_MAX_SEGMENTS) {
 		pr_warn("Clamping indirect_sg_entries to %u\n",
 			SG_MAX_SEGMENTS);
 		indirect_sg_entries = SG_MAX_SEGMENTS;
 	}
+#else
+	if (indirect_sg_entries > SCSI_MAX_SG_CHAIN_SEGMENTS) {
+		pr_warn("Clamping indirect_sg_entries to %u\n",
+			SCSI_MAX_SG_CHAIN_SEGMENTS);
+		indirect_sg_entries = SCSI_MAX_SG_CHAIN_SEGMENTS;
+	}
+#endif
 
 	srp_remove_wq = create_workqueue("srp_remove");
 	if (!srp_remove_wq) {
--- a/drivers/infiniband/ulp/srp/ib_srp.h
+++ b/drivers/infiniband/ulp/srp/ib_srp.h
@@ -80,6 +80,23 @@ enum srp_iu_type {
 	SRP_IU_RSP,
 };
 
+#ifndef HAVE_BLK_MQ_UNIQUE_TAG
+static inline u32 build_srp_tag(u16 ch, u16 req_idx)
+{
+	return ch << 16 | req_idx;
+}
+
+static inline u16 srp_tag_ch(u32 tag)
+{
+	return tag >> 16;
+}
+
+static inline u16 srp_tag_idx(u32 tag)
+{
+	return tag & ((1 << 16) - 1);
+}
+#endif
+
 /*
  * @mr_page_mask: HCA memory registration page mask.
  * @mr_page_size: HCA memory registration page size.
@@ -112,6 +129,9 @@ struct srp_host {
 };
 
 struct srp_request {
+#ifndef HAVE_BLK_MQ_UNIQUE_TAG
+	struct list_head        list;
+#endif
 	struct scsi_cmnd       *scmnd;
 	struct srp_iu	       *cmd;
 	union {
@@ -122,6 +142,9 @@ struct srp_request {
 	struct srp_direct_buf  *indirect_desc;
 	dma_addr_t		indirect_dma_addr;
 	short			nmdesc;
+#ifndef HAVE_BLK_MQ_UNIQUE_TAG
+	uint32_t		tag;
+#endif
 	struct ib_cqe		reg_cqe;
 };
 
@@ -132,6 +155,9 @@ struct srp_request {
 struct srp_rdma_ch {
 	/* These are RW in the hot path, and commonly used together */
 	struct list_head	free_tx;
+#ifndef HAVE_BLK_MQ_UNIQUE_TAG
+	struct list_head        free_reqs;
+#endif
 	spinlock_t		lock;
 	s32			req_lim;
 
@@ -181,6 +207,9 @@ struct srp_target_port {
 	/* read only in the hot path */
 	struct ib_pd		*pd;
 	struct srp_rdma_ch	*ch;
+#ifndef HAVE_BLK_MQ_UNIQUE_TAG
+	int			*mq_map;
+#endif
 	u32			ch_count;
 	u32			lkey;
 	enum srp_target_state	state;
--- a/drivers/scsi/scsi_priv.h
+++ b/drivers/scsi/scsi_priv.h
@@ -188,5 +188,10 @@ static inline void scsi_dh_remove_device
  */
 
 #define SCSI_DEVICE_BLOCK_MAX_TIMEOUT	600	/* units in seconds */
+#ifndef HAVE_SCSI_DEVICE_SCSI_INTERNAL_DEVICE_BLOCK
+extern int scsi_internal_device_block(struct scsi_device *sdev);
+extern int scsi_internal_device_unblock(struct scsi_device *sdev,
+					enum scsi_device_state new_state);
+#endif
 
 #endif /* _SCSI_PRIV_H */
--- a/drivers/scsi/scsi_transport_srp.c
+++ b/drivers/scsi/scsi_transport_srp.c
@@ -19,6 +19,7 @@
  * 02110-1301 USA
  */
 #include <linux/init.h>
+#include <linux/version.h>
 #include <linux/module.h>
 #include <linux/jiffies.h>
 #include <linux/err.h>
@@ -411,7 +412,11 @@ static void __rport_fail_io_fast(struct
 	 * calls before invoking i->f->terminate_rport_io().
 	 */
 	scsi_target_block(rport->dev.parent);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 6, 0) || defined(CONFIG_COMPAT_SCSI_TARGET_UNBLOCK)
 	scsi_target_unblock(rport->dev.parent, SDEV_TRANSPORT_OFFLINE);
+#else
+	scsi_target_unblock(rport->dev.parent);
+#endif
 
 	/* Involve the LLD if possible to terminate all I/O on the rport. */
 	i = to_srp_internal(shost->transportt);
@@ -454,7 +459,11 @@ static void rport_dev_loss_timedout(stru
 
 	mutex_lock(&rport->mutex);
 	WARN_ON(srp_rport_set_state(rport, SRP_RPORT_LOST) != 0);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 6, 0) || defined(CONFIG_COMPAT_SCSI_TARGET_UNBLOCK)
 	scsi_target_unblock(rport->dev.parent, SDEV_TRANSPORT_OFFLINE);
+#else
+	scsi_target_unblock(rport->dev.parent);
+#endif
 	mutex_unlock(&rport->mutex);
 
 	i->f->rport_delete(rport);
@@ -553,7 +562,11 @@ int srp_reconnect_rport(struct srp_rport
 
 		rport->failed_reconnects = 0;
 		srp_rport_set_state(rport, SRP_RPORT_RUNNING);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 6, 0) || defined(CONFIG_COMPAT_SCSI_TARGET_UNBLOCK)
 		scsi_target_unblock(&shost->shost_gendev, SDEV_RUNNING);
+#else
+		scsi_target_unblock(&shost->shost_gendev);
+#endif
 		/*
 		 * If the SCSI error handler has offlined one or more devices,
 		 * invoking scsi_target_unblock() won't change the state of
@@ -571,12 +584,20 @@ int srp_reconnect_rport(struct srp_rport
 		 * failure timers if these had not yet been started.
 		 */
 		__rport_fail_io_fast(rport);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 6, 0) || defined(CONFIG_COMPAT_SCSI_TARGET_UNBLOCK)
 		scsi_target_unblock(&shost->shost_gendev,
 				    SDEV_TRANSPORT_OFFLINE);
+#else
+		scsi_target_unblock(&shost->shost_gendev);
+#endif
 		__srp_start_tl_fail_timers(rport);
 	} else if (rport->state != SRP_RPORT_BLOCKED) {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 6, 0) || defined(CONFIG_COMPAT_SCSI_TARGET_UNBLOCK)
 		scsi_target_unblock(&shost->shost_gendev,
 				    SDEV_TRANSPORT_OFFLINE);
+#else
+		scsi_target_unblock(&shost->shost_gendev);
+#endif
 	}
 	mutex_unlock(&rport->mutex);
 
